### PR TITLE
ESI-13714 Added Throw function to OTTOmock and custom JS Exception.

### PR DESCRIPTION
**Pull request by andriikonstantinov**

### CODE REVIEW

File: vendor/github.com/coreos/etcd/clientv3/client.go [hunk 1]
```diff
@@ -144,7 +144,7 @@ func (c *Client) autoSync() {
 		case <-c.ctx.Done():
 			return
 		case <-time.After(c.cfg.AutoSyncInterval):
-			ctx, _ := context.WithTimeout(c.ctx, 5*time.Second)
+			ctx, _ := context.WithTimeout(c.ctx, 5*time.Millisecond)
```
Comment (p-kozlowski) [inline vendor/github.com/coreos/etcd/clientv3/client.go]: what's the rationale of this change? could you link a PR to etcd?

File: extension/framework/runner/environment.go [hunk 2]
```diff
@@ -277,14 +276,21 @@ func (env *Environment) mockFunction(functionName string) {
 				functionName, readableArguments, valueSliceToString(expectedArguments)))
 		}
 
-		response := responses[0]
-		responses = responses[1:]
-		env.setToOtto(functionName, "responses", responses)
+		var answer otto.Value
+		if len(exceptions) > 0 {
```
Comment (p-kozlowski) [inline extension/framework/runner/environment.go]: this won't work. consider:
```
someCall().Expect().Return(1)
someCall().Expect().Return(2)
someCall().Expect().Throw(3)
```
The calls should be processed in order, currently the first call will throw.

File: extension/framework/runner/test_data/gohan_sync_fetch_mock.js [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+var PATH = "/v1.0/networks";
+
+function testUnexpectedCall() {
+    result = gohan_sync_fetch("unexpectedKey");
+}
+
+function testSingleReturnSingleCall() {
+    var mock_response = "TimeOutException";
+    gohan_sync_fetch.Expect("timeOutKey").Throw(mock_response);
+    var response = gohan_sync_fetch("timeOutKey");
```
Comment (p-kozlowski) [inline extension/framework/runner/test_data/gohan_sync_fetch_mock.js]: this should throw - check if the proper exception was caught

File: extension/framework/runner/test_data/gohan_sync_fetch_mock.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    if (response !== mock_response) {
```
Comment (p-kozlowski) [inline extension/framework/runner/test_data/gohan_sync_fetch_mock.js]: this comparison doesn't make sense - firstly, it's a dead code, secondly, function that throws can't return anything (same elsewhere)

File: extension/framework/runner/environment.go [hunk 5]
```diff
@@ -300,11 +302,28 @@ func (env *Environment) mockFunction(functionName string) {
...
 		}
 		responses = append(responses, call.ArgumentList[0])
+		exceptions = append(exceptions, otto.FalseValue())
+		env.setToOtto(functionName, "responses", responses)
+		env.setToOtto(functionName, "exceptions", exceptions)
+		return otto.NullValue()
+	})
+
+	env.setToOtto(functionName, "exceptions", []otto.Value{})
+	env.setToOtto(functionName, "Throw", func(call otto.FunctionCall) otto.Value {
```
Comment (p-kozlowski) [inline extension/framework/runner/environment.go]: code duplication: Throw is (almost) exactly the same as Return

File: extension/framework/runner/runner_test.go [hunk 6]
```diff
@@ -349,6 +349,20 @@ var _ = Describe("Runner", func() {
 			})
 		})
 
+		Context("When correctly using the Gohan Sync Fetch mock", func() {
+			BeforeEach(func() {
+				testFile = "./test_data/gohan_sync_fetch_mock.js"
+			})
+
+			FIt("Should work", func() {
```
Comment (p-kozlowski) [inline extension/framework/runner/runner_test.go]: don't push Focused test cases

File: extension/framework/runner/test_data/gohan_sync_fetch_mock.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+    try {
+        gohan_sync_fetch()
+    } catch(error) {
+        if (error instanceof TimeoutException){
+            return;
+        }else{
+            Fail("unexpected Exception")
+        }
+    }
```
Comment (p-kozlowski) [inline extension/framework/runner/test_data/gohan_sync_fetch_mock.js]: you should fail if the exception was not thrown at all

File: extension/framework/runner/test_data/gohan_sync_fetch_mock.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+function testWrongThrow() {
+    gohan_sync_fetch.Expect().Throw("FakeException");
+    try {
+        gohan_sync_fetch()
+    } catch(error) {
+        if (error instanceof TimeoutException){
```
Comment (p-kozlowski) [inline extension/framework/runner/test_data/gohan_sync_fetch_mock.js]: I'd rather use "positive" test cases, example

```JS
function testShouldThrowCorrectExceptionType {
  gohan_sync_fetch.Expect().Throw("TimeoutException");
  try {
    gohan_sync_fetch()
    Fail("exception not thrown");
  } catch(error) {
    if (error instanceof TimeoutException){
      return; // everything's fine
    } else {
      Fail("wrong exception type");
    }
  }
}
```

File: extension/otto/otto.go [hunk 9]
```diff
@@ -354,14 +354,17 @@ func (env *Environment) ClearEnvironment() {
...
 	panic(exception)
 }
 
 //ThrowOttoException throws a JavaScript exception that will be passed to Otto
 func ThrowOttoException(call *otto.FunctionCall, format string, arguments ...interface{}) {
-	throwOtto(call, "Error", fmt.Sprintf(format, arguments...))
+	ThrowOtto(call, "Error", fmt.Sprintf(format, arguments...))
+}
+func ThrowTimeOutException(call *otto.FunctionCall, format string, arguments ...interface{}){
+	ThrowOtto(call, "TimeOutException", fmt.Sprintf(format, arguments...))
```
Comment (p-kozlowski) [inline extension/otto/otto.go]: TimeOutException should be defined in gohan_core.go (cf. CustomException, ResourceException, etc.)

**Comment by przemyslaw-dobrowolski-cl** [comment]
@andriikonstantinov please rebase because there is a conflict

File: cli/cli.coverprofile [hunk 10]
```diff
@@ -0,0 +1,265 @@
+mode: atomic
+github.com/cloudwan/gohan/cli/cli.go:55.39,83.2 8 0
```
Comment (przemyslaw-dobrowolski-cl) [inline cli/cli.coverprofile]: why this a part of this PR??

**Comment by przemyslaw-dobrowolski-cl** [comment]
@andriikonstantinov please resolve the conflict

File: docs/configuration.md [hunk 11]
```diff
@@ -48,6 +48,8 @@ Gohan server configuration uses YAML format.
   # list of etcd backend servers
   etcd:
       - "http://127.0.0.1:2379"
+  # timeout in milliseconds.1000 milliseconds by default
```
Comment (przemyslaw-dobrowolski-cl) [inline docs/configuration.md]: in line 51 you write that 1000 ms is the default
in line 52 you use 333 ms. It is misleading.

File: extension/framework/runner/environment.go [hunk 12]
```diff
@@ -298,16 +313,10 @@ func (env *Environment) mockFunction(functionName string) {
...
-			call.Otto.Call("Fail", nil, "Return() should be called with exactly one argument")
-		}
-		responses = append(responses, call.ArgumentList[0])
-		env.setToOtto(functionName, "responses", responses)
+	env.setToOtto(functionName, "Return", getSpecifiedFunction(env,functionName,otto.FalseValue()))
 
-		return otto.NullValue()
-	})
+	env.setToOtto(functionName, "exceptions", []otto.Value{})
+	env.setToOtto(functionName, "Throw", getSpecifiedFunction(env,functionName,otto.TrueValue()))
```
Comment (przemyslaw-dobrowolski-cl) [inline extension/framework/runner/environment.go]: In one place there is a space after command and in other not. Fix the inconsistency.

File: extension/framework/runner/runner_test.go [hunk 13]
```diff
@@ -349,6 +349,20 @@ var _ = Describe("Runner", func() {
 			})
 		})
 
+		Context("When correctly using the Gohan Sync Fetch mock", func() {
+			BeforeEach(func() {
+				testFile = "./test_data/gohan_sync_fetch_mock.js"
+			})
+
+			It("Should work", func() {
+				Expect(errors).To(HaveLen(3))
```
Comment (przemyslaw-dobrowolski-cl) [inline extension/framework/runner/runner_test.go]: Where is errors defined? Why we check it at the beginning of the test?

File: extension/framework/runner/test_data/gohan_sync_fetch_mock.js [hunk 14]
```diff
@@ -0,0 +1,63 @@
+// Copyright (C) 2015 NTT Innovation Institute, Inc.
```
Comment (przemyslaw-dobrowolski-cl) [inline extension/framework/runner/test_data/gohan_sync_fetch_mock.js]: As far as I know, it's 2017 now.

File: extension/otto/gohan_core.go [hunk 15]
```diff
@@ -109,6 +109,23 @@ func init() {
 		    return this.name.concat("(").concat(this.message).concat(")");
 		  };
 		}
+		function TimeOutException(msg){
+		  BaseException.call(this);
+		  this.message = msg;
+		  this.name = "TimeOutException"
+		  this.code = 408
+    	  this.fields.push("code")
```
Comment (przemyslaw-dobrowolski-cl) [inline extension/otto/gohan_core.go]: wrong indentation

File: sync/util/etcd.go [hunk 16]
```diff
@@ -26,7 +26,7 @@ func CreateFromConfig(config *util.Config) (s sync.Sync, err error) {
 		etcdServers := config.GetStringList("etcd", nil)
 		if etcdServers != nil {
 			log.Info("etcd servers: %s", etcdServers)
-			s, err = etcdv3.NewSync(etcdServers, time.Second)
+			s, err = etcdv3.NewSync(etcdServers,  time.Duration(config.GetInt("etcd_timeout",1000)) * time.Millisecond)
```
Comment (przemyslaw-dobrowolski-cl) [inline sync/util/etcd.go]: Define 1000 as a constant with some particular meaning

File: extension/framework/runner/runner_test.go (hunk 13 above)
Comment (andriikonstantinov) [inline extension/framework/runner/runner_test.go]: @przemyslaw-dobrowolski-cl The previous tests suggest such implementation.

File: extension/framework/runner/environment.go [hunk 17]
```diff
@@ -254,19 +254,31 @@ func (env *Environment) getTransaction(isNew bool, options *transaction.TxOption
 	env.dbTransactions = append(env.dbTransactions, tx)
 	return tx, nil
 }
-
+func getSpecifiedFunction(env *Environment, functionName string, setValue otto.Value) func(call otto.FunctionCall) otto.Value {
+	return func(call otto.FunctionCall) otto.Value {
+		exceptions := env.getFromOtto(functionName, "exceptions").([]otto.Value)
+		responses := env.getFromOtto(functionName, "responses").([]otto.Value)
+		if len(call.ArgumentList) != 1 {
+			call.Otto.Call("Fail", nil, "Return() should be called with exactly one argument")
```
Comment (p-kozlowski) [inline extension/framework/runner/environment.go]: misleading message: it does not have to be Return()

File: extension/otto/gohan_core.go [hunk 18]
```diff
@@ -109,6 +109,23 @@ func init() {
 		    return this.name.concat("(").concat(this.message).concat(")");
 		  };
 		}
+		function TimeOutException(msg){
+		  BaseException.call(this);
+		  this.message = msg;
+		  this.name = "TimeOutException";
+		  this.code = 408;
+    	  this.fields.push("code");
```
Comment (p-kozlowski) [inline extension/otto/gohan_core.go]: indent

File: extension/otto/gohan_sync.go [hunk 19]
```diff
@@ -153,6 +156,10 @@ func init() {
 					ThrowOttoException(&call, "Invalid type of second argument: expected a string")
 					return otto.NullValue()
 				}
+				if value, err = GetString(call.Argument(1)); err != nil {
```
Comment (p-kozlowski) [inline extension/otto/gohan_sync.go]: why have you copy-pasted the above 4 lines?

File: extension/otto/gohan_sync.go [hunk 20]
```diff
@@ -172,6 +179,24 @@ func init() {
 					}
 				}
 
+				errCh = make(chan error)
```
Comment (p-kozlowski) [inline extension/otto/gohan_sync.go]: again a copy-paste

File: sync/util/etcd.go [hunk 21]
```diff
@@ -27,7 +27,8 @@ func CreateFromConfig(config *util.Config) (s sync.Sync, err error) {
 		etcdServers := config.GetStringList("etcd", nil)
 		if etcdServers != nil {
 			log.Info("etcd servers: %s", etcdServers)
-			s, err = etcdv3.NewSync(etcdServers, time.Second)
+			etcdTimeoutDefaultValue := 1000
```
Comment (p-kozlowski) [inline sync/util/etcd.go]: etcdTimeoutDefaultValue should be a const

File: docs/configuration.md (hunk 11 above)
Comment (przemyslaw-dobrowolski-cl) [inline docs/configuration.md]: this was not addressed

File: extension/framework/runner/runner_test.go (hunk 13 above)
Comment (przemyslaw-dobrowolski-cl) [inline extension/framework/runner/runner_test.go]: ok

File: extension/framework/runner/test_data/gohan_sync_fetch_mock.js (hunk 14 above)
Comment (przemyslaw-dobrowolski-cl) [inline extension/framework/runner/test_data/gohan_sync_fetch_mock.js]: this was not adressed

File: extension/framework/runner/test_data/gohan_sync_fetch_mock.js [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+        gohan_sync_fetch()
+    } catch(error) {
+        if (error instanceof TimeOutException){
+            return;
+        }
+    }
+    Fail()
+}
+
+
```
Comment (przemyslaw-dobrowolski-cl) [inline extension/framework/runner/test_data/gohan_sync_fetch_mock.js]: why so many enters here?

File: docs/configuration.md (hunk 11 above)
Comment (andriikonstantinov) [inline docs/configuration.md]: @przemyslaw-dobrowolski-cl fixed all the comments.

File: extension/otto/gohan_sync.go [hunk 23]
```diff
@@ -86,7 +86,10 @@ func init() {
 					log.Debug("Received otto interrupt in gohan_sync_fetch")
 					interrupt()
 				case err = <-errCh:
-					if err != nil {
+					if err != nil && err == context.DeadlineExceeded {
```
Comment (marcin-ptaszynski) [inline extension/otto/gohan_sync.go]: `err == context.DeadlineExceeded` already implicates `err != nil`

File: sync/etcdv3/etcd_test.go [hunk 24]
```diff
@@ -1,12 +1,11 @@
 package etcdv3
 
 import (
-	"testing"
-	"time"
-
 	gohan_sync "github.com/cloudwan/gohan/sync"
 	etcd "github.com/coreos/etcd/clientv3"
 	"golang.org/x/net/context"
+	"testing"
```
Comment (marcin-ptaszynski) [inline sync/etcdv3/etcd_test.go]: broken ordering. stdlib should be in prior import group.

File: sync/util/etcd.go [hunk 25]
```diff
@@ -13,6 +13,8 @@ import (
 
 var log = l.NewLogger()
 
+const etcdTimeoutDefaultValue = 1000
```
Comment (marcin-ptaszynski) [inline sync/util/etcd.go]: if type is not `time.Duration`, please denote it's `ms` with name suffix.

File: docs/configuration.md [hunk 26]
```diff
@@ -52,6 +52,8 @@ Gohan server configuration uses YAML format.
   # list of etcd backend servers
   etcd:
       - "http://127.0.0.1:2379"
+  # timeout in milliseconds.1000 milliseconds by default 
+  etcd_timeout: 1000
```
Comment (marcin-ptaszynski) [inline docs/configuration.md]: make units clear with `_ms` suffix?

Comment (andriikonstantinov) [inline docs/configuration.md]: Done

**Comment by andriikonstantinov** [comment]
@p-kozlowski @przemyslaw-dobrowolski-cl @marcin-ptaszynski please review.

File: extension/framework/runner/test_data/gohan_sync_fetch_mock.js [hunk 27]
```diff
@@ -0,0 +1,60 @@
...
+// implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+var SCHEMA_INCLUDES = [];
+var SCHEMAS = ["./schema.yaml"];
+var PATH = "/v1.0/networks";
+
+function testUnexpectedCall() {
+    result = gohan_sync_fetch("unexpectedKey");
```
Comment (p-kozlowski) [inline extension/framework/runner/test_data/gohan_sync_fetch_mock.js]: "result" becomes a global variable. change to `var result = ...` or simply `gohan_sync_fetch("unexpected")`

File: extension/otto/otto.go [hunk 28]
```diff
@@ -354,14 +354,15 @@ func (env *Environment) ClearEnvironment() {
 	env.DataStore.Close()
 }
 
-func throwOtto(call *otto.FunctionCall, exceptionName string, arguments ...interface{}) {
+//ThrowOtto throws cutsom JavaScript exception that will be passed to Otto
```
Comment (p-kozlowski) [inline extension/otto/otto.go]: typo: custom

**Comment by andriikonstantinov** [comment]
@nati please review.

**Comment by andriikonstantinov** [comment]
@morrisson , @nati please review

File: extension/framework/runner/environment.go [hunk 29]
```diff
@@ -254,19 +254,33 @@ func (env *Environment) getTransaction(isNew bool, options *transaction.TxOption
 	env.dbTransactions = append(env.dbTransactions, tx)
 	return tx, nil
 }
-
+func getSpecifiedFunction(env *Environment, functionName string, setValue otto.Value) func(call otto.FunctionCall) otto.Value {
```
Comment (morrisson) [inline extension/framework/runner/environment.go]: nit: should include white space between functions.

File: etc/gohan.yaml [hunk 30]
```diff
@@ -37,8 +37,10 @@ tls:
 # embedded webui get served if you use "embed" here
 document_root: "embed"
 # list of etcd backend servers
-#etcd:
-#    - "http://127.0.0.1:2379"
+etcd:
+    - "http://127.0.0.1:2379"
+
+etcd_timeout_ms: 1000
```
Comment (morrisson) [inline etc/gohan.yaml]: Should include comment for `etcd_timeout_ms`

File: extension/otto/gohan_core.go [hunk 31]
```diff
@@ -109,6 +109,23 @@ func init() {
 		    return this.name.concat("(").concat(this.message).concat(")");
 		  };
 		}
+		function TimeOutException(msg){
+		  BaseException.call(this);
+		  this.message = msg;
+		  this.name = "TimeOutException";
+		  this.code = 408;
```
Comment (morrisson) [inline extension/otto/gohan_core.go]: This error happens based on server side overload, so it should return 5xx server error. That should be 503 Service Unavailable. 408 means "The server timed out waiting for the request" so it's different from etcd timeout case.

File: extension/framework/runner/environment.go (hunk 29 above)
Comment (andriikonstantinov) [reply extension/framework/runner/environment.go]: Added

File: etc/gohan.yaml (hunk 30 above)
Comment (andriikonstantinov) [reply etc/gohan.yaml]: Done

**Comment by andriikonstantinov** [comment]
@morrisson please, take a look.
